20231214-listUsers-api.md

```diff
@@ -0,0 +1,887 @@
+| document:1 | viewer   | user:andres |
+
+
+```
+ListUsers({
+  object: "document:1",
+  relation: "viewer",
+  user_filter: [{type: "user"}]
+}) --> {
+  users: ["user:jon", "user:andres"],
+  wildcard_types: []
+}
+```
+In this case the algorithm simply must expand all direct relationships between `document:1#viewer` and `user` objects. Since these are directly related to one another, then we must only do a simple reverse database lookup.
+
+### Example 1b (Direct Relation with Userset)
+```
+model
+  schema 1.1
+
+type user
+
+type group
+  relations
+    define member: [user, group#member]
+
+type document
+  relations
+    define viewer: [group#member]
+```
+| object         | relation | user                  |
+|----------------|----------|-----------------------|
+| document:1     | viewer   | group:eng#member      |
+| group:eng      | member   | group:fga#member      |
+| group:fga      | member   | user:andres           |
+| group:fga      | member   | group:fga-core#member |
+| group:fga-core | member   | user:jon              |
+
+```
+ListUsers({
+  object: "document:1",
+  relation: "viewer",
+  user_filter: [{type: "user"}]
+}) --> ["user:jon", "user:andres"]
+```
+This example requires recursive expansion.
+> ℹ️ The usage of `expand` from here forward refers to an internal expand function, and it should not be confused with the public Expand API, but it operates somewhat similarly.
+
+
+1. Expand the set of subjects/users that relate to `document:1#viewer`. 
+
+   `expand(document:1#viewer)` --> ["group:eng#member"]
+
+1. Expand the new set of subjects/users that make up `group:eng#member` set.
+
+   `expand(group:eng#member)` --> ["group:fga#member"]
+
+1. Expand the set of subjects/users that make up the `group:fga#member` set.
+
+   `expand(group:fga#member)` --> ["user:andres", "group:fga-core#member"]
+
+   1. We find a terminal/concrete object of `user:andres`, which matches the `user` in the user filter, and so we add it to the list of items to include in the response.
+
+1. Continue expanding the residual set of subjects/users in `group:fga-core#member`
+
+   `expand(group:fga-core#member)` --> ["user:jon"]
+
+   1. We find a terminal/concrete object of `user:jon`, which matches the `user` in the user filter, so add it to the list of items to include in the response
+
+1. No further subjects to expand, and so we're done. Return the items we accumulated from the steps above.
+
+Visually, the overall recursive call tree looks like the following:
+
+```
+expand(document:1#viewer) --> ["group:eng#member"]
+|-> expand(group:eng#member) --> ["group:fga#member"]
+|---> expand(group:fga#member) --> ["user:andres", "group:fga-core#member"]
+      yield "user:andres"
+|-----> expand(group:fga-core#member) --> ["user:jon"]
+        yield "user:jon"
+
+return ["user:andres", "user:jon"]
+```
+
+### Example 1c (Direct Relation with Typed Public Wildcard)
+See [Typed Public Wildcard](#typed-public-wildcard) section for more context.
```

Comment:
Nitpick, but we call this `Type Bound Public Access`, we try to shy away from the term wildcard as it indicates regex-like behavior

Reply:
The concept is referring to Type Bound Public Access. The type definition / structure itself is defined as `TypedPublicWildcard`. Not sure how we want to delineate that, but our concepts should match our API protobuf definitions ideally.

Reply:
The API uses `wildcard` as the field definition in existing protobuf definitions as well. So the ship has sailed a bit on using the term "wildcard"